### PR TITLE
Migrate the ExpertSearch component

### PR DESCRIPTION
**Pull request by flavens**
resolves [DSP-390](https://dasch.myjetbrains.com/youtrack/issue/DSP-390)
closes [DSP-359](https://dasch.myjetbrains.com/youtrack/issue/DSP-359)

Include the update of the main dsp-ui README ([DSP-395](https://dasch.myjetbrains.com/youtrack/issue/DSP-395)).

Currently using "@dasch-swiss/dsp-js": "1.0.0-rc.4" and "dsp-api": "13.0.0-rc.4"

### CODE REVIEW

**Comment by flavens** [comment]
@tobiasschweizer Does it make sense to create e2e tests for expert search? If yes, which kind of tests do we need?

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.scss [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+.form-content {
+  margin: 24px auto;
+  width: 472px;
+
+  .large-field {
+    min-width: 472px;
+  }
+}
```
Comment (kilchenmann) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.scss:32]: Can you add a TODO comment here? The general form settings should degined in a main stylesheet in the module's root folder. But this should be part of a separate PR.

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * @param gravsearchQuery Send the gravsearch query back.
+     */
+    @Output() gravsearchQuery = new EventEmitter<string>();
+
+    /**
+     * @param toggleExtendedSearchForm Trigger toggle for extended search form.
+     */
+    @Output() toggleExpertSearchForm = new EventEmitter<boolean>();
```
Comment (kilchenmann) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:23]: `toggleExtendedSearchForm` should be `toggleExpertSearchForm`

File: src/app/search-playground/search-playground.component.scss [hunk 3]
```diff
@@ -0,0 +1,8 @@
+.expert-search-container {
+  min-height: 100%;
+  min-width: 100%;
+}
+
+.mat-input-element {
+  font-family: 'Courier New', Courier, monospace
+}
```
Comment (kilchenmann) [inline src/app/search-playground/search-playground.component.scss:8]: Same as in [ projects/dsp-ui/src/lib/search/expert-search/expert-search.component.scss](https://github.com/dasch-swiss/dsp-ui-lib/pull/114/files#diff-85ff81ac3b3179331a4d85ae475d026f) line 21-23
It doesn't need the same style here again?

Comment (flavens) [reply src/app/search-playground/search-playground.component.scss]: done in 0e81d2c

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 2 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: done in 59c1d2d

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.scss (hunk 1 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.scss]: done in 59c1d2d

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    expertSearchForm: FormGroup;
+
+    constructor(
+        @Inject(DspApiConfigToken) private dspApiConfig: KnoraApiConfig,
+        private _searchParamsService: AdvancedSearchParamsService,
+        private fb: FormBuilder,
+        private _route: ActivatedRoute,
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:31]: Since you made an `@Output` (which is good) I think you can remove the route and the router.

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    submitQuery() {
+        const gravsearch = this.generateGravsearch(0);
+
+        if (gravsearch) {
+            console.log('submitQuery gravsearch', gravsearch);
+            this.gravsearchQuery.emit(gravsearch);
+        }
+
+        // toggle expert search form
+        this.toggleExpertSearchForm.emit(true);
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:84]: Do you need an extra `@Output` for the toggle? Or could you just toggle on emission of a Gravsearch query?

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+}
+
+describe('ExpertSearchComponent', () => {
+  let testHostComponent: TestHostComponent;
+  let testHostFixture: ComponentFixture<TestHostComponent>;
+  let hostCompDe: DebugElement;
+
+  const dspConfSpy = new KnoraApiConfig('http', 'localhost', 3333, undefined, undefined, true);
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts:35]: Move this into the `beforeEach`. I think here it does not make a difference, but it is a good practice.

If there are side-effects (setting a token etc.), then this could lead to unpredictable behaviour since the `it`s are run in random order.

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    /**
+     * @ignore
+     * Initiate the form with predefined Gravsearch query as example.
+     */
+    private initForm() {
+        this.expertSearchForm = this.fb.group({
+            gravquery: [
+`PREFIX knora-api: <http://api.knora.org/ontology/knora-api/simple/v2#>
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:46]: please assign the query string to a variable and then use the variable here.

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    ?book incunabula:title ?title .
+
+} WHERE {
+    ?book a incunabula:book .
+    ?book incunabula:title ?title .
+}
+`
+    );
+  });
+
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts:133]: you should add an  `it` that checks for the correct registration of the query in the params service.

I think this is identical to what I did for the Gravsearch generation service:

https://github.com/dasch-swiss/dsp-ui-lib/blob/b3b1fee1ecc97085f3bdb199c0352c36825b206a/projects/dsp-ui/src/lib/search/services/gravsearch-generation.service.spec.ts#L14-L31

https://github.com/dasch-swiss/dsp-ui-lib/blob/b3b1fee1ecc97085f3bdb199c0352c36825b206a/projects/dsp-ui/src/lib/search/services/gravsearch-generation.service.spec.ts#L78-L80

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+    ngOnInit(): void {
+        this.initForm();
+    }
+
+    /**
+     * @ignore
+     * Initiate the form with predefined Gravsearch query as example.
+     */
+    private initForm() {
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:43]: You can also instantiate a `FormControl` and assign to a class member:

https://github.com/dasch-swiss/dsp-ui-lib/blob/b3b1fee1ecc97085f3bdb199c0352c36825b206a/projects/dsp-ui/src/lib/viewer/values/int-value/int-value.component.ts#L41-L55

Then you do not need to access it via the string within your class.

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * @ignore
+     * Generate the whole gravsearch query matching the query given by the form.
+     */
+    private generateGravsearch(offset: number = 0): string {
+        const queryTemplate = this.expertSearchForm.controls['gravquery'].value;
+
+        // offset component of the Gravsearch query
+        const offsetTemplate = `
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:95]: What if the user already provided an OFFSET?

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+import { KnoraApiConfig } from '@dasch-swiss/dsp-js';
+import { DspApiConfigToken } from '../../core/core.module';
+import { ExpertSearchComponent } from './expert-search.component';
+
+/**
+ * Test host component to simulate parent component.
+ */
+@Component({
+    template: `
+        <dsp-expert-search #expSearch></dsp-expert-search>`
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts:19]: In the test host component you should register to the `@Output`s. Then you can trigger the emission and see if it had the desired effect. 

https://github.com/dasch-swiss/dsp-ui-lib/blob/b3b1fee1ecc97085f3bdb199c0352c36825b206a/projects/dsp-ui/src/lib/search/advanced-search/select-ontology/select-ontology.component.spec.ts#L14-L44

https://github.com/dasch-swiss/dsp-ui-lib/blob/b3b1fee1ecc97085f3bdb199c0352c36825b206a/projects/dsp-ui/src/lib/search/advanced-search/select-ontology/select-ontology.component.spec.ts#L117-L133

File: projects/dsp-ui/design-documentation.md [hunk 12]
```diff
@@ -155,7 +155,7 @@ The full text search quickly finds all instances of a term in DSP-API. It can be
 
 ### Expert search
 
-The expert search is a textarea which allows you to manually write Gravsearch queries if you are comfortable with the language and want more precise results.
+The expert search is a textarea which allows you to manually write Gravsearch queries if you are comfortable with the language and want more precise results. The textarea content contains a default Gravsearch query as an example for the user. It is possible to reset the field, back to the default query.
```
Comment (tobiasschweizer) [inline projects/dsp-ui/design-documentation.md:158]: ", back to the default query" sounds weird. @mdelez?

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+CONSTRUCT {
+    ?book knora-api:isMainResource true .
+    ?book incunabula:title ?title .
+
+} WHERE {
+    ?book a incunabula:book .
+    ?book incunabula:title ?title .
+}
+`,
+                Validators.required
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:58]: There should be a validator that checks for a user-provided offset.

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 4 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: done in b7417bc

File: projects/dsp-ui/design-documentation.md (hunk 12 above)
Comment (mdelez) [reply projects/dsp-ui/design-documentation.md:158]: If you're talking about the comma, it should be removed. The sentence is fine otherwise.

Comment (flavens) [reply projects/dsp-ui/design-documentation.md]: done in f6128fb

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts (hunk 11 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts]: done in 062ccea

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts (hunk 8 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts:155]: done in 062ccea

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts (hunk 6 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts]: done in 062ccea

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 5 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: We needed it in knora-ui because of the expansion-panel we used. But this has not been implemented in extended-search, so I removed it here. If we want to implement the same design, we will do it in a different PR.

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 9 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: done in 0fe3539

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 7 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:23]: done in ea9edea

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 13 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: done in 79f3143 + bbc3af4

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 10 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:98]: done in 79f3143 + bbc3af4

**Comment by flavens** [comment]
> In general, I think this component would be much more useful if
> 
> * it provided syntax checking (e..g, https://github.com/RubenVerborgh/SPARQL.js/)
> * it added support for prefixes of Knora ontologies and supported standard ontologies
> * it did not allow the user to add an `OFFSET`

- it provided syntax checking -> reported in [DSP-493](https://dasch.myjetbrains.com/youtrack/issue/DSP-493)
- it added support for prefixes [...] -> reported in [DSP-492](https://dasch.myjetbrains.com/youtrack/issue/DSP-492)
- it did not allow the user to add an `OFFSET` -> fixed in the PR

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 4 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: ok

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 5 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: ok

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts (hunk 6 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts]: ok

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 7 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: ok

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts (hunk 8 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts:260]: ok

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 9 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: You can now use `this.queryFormControl.value` instead of `const queryTemplate = this.expertSearchForm.controls['gravsearchquery'].value;`

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 10 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:98]: ok

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts (hunk 11 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.spec.ts]: ok, great

File: projects/dsp-ui/design-documentation.md (hunk 12 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/design-documentation.md]: yes, the comma confused me :-)

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 13 above)
Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: ok

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.html [hunk 14]
```diff
@@ -0,0 +1,28 @@
+<div class="expert-search-container">
+    <form [formGroup]="expertSearchForm" class="expert-search-form kui-form-content">
+        <mat-form-field class="textarea-field large-field">
+            <textarea matInput [formControlName]="'gravsearchquery'" cdkTextareaAutosize cdkAutosizeMinRows="12"
+                cdkAutosizeMaxRows="24" class="textarea-field-content" placeholder="Write your gravsearch query">
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.html:5]: Gravsearch -> upper case

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 9 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts]: done in 26e2af2

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        if (gravsearch) {
+            this.gravsearchQuery.emit(gravsearch);
+        }
+    }
+
+    /**
+     * @ignore
+     * Generate the whole gravsearch query matching the query given by the form.
+     */
+    private _generateGravsearch(offset: number = 0): string {
```
Comment (tobiasschweizer) [inline projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:94]: I think you can assume that the offset is always 0 since the user is not allowed to provide an offset.

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.html (hunk 14 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.html]: done in 5b6277b

File: projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts (hunk 15 above)
Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:94]: so I can remove `= 0`?

Comment (tobiasschweizer) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:94]: Please just create a DSP task so we can check this in

- expert-search
- gravsearch-generation-service

I think I should so that after my holiday ;-)

Comment (flavens) [reply projects/dsp-ui/src/lib/search/expert-search/expert-search.component.ts:94]: discussion moved to [DSP-499](https://dasch.myjetbrains.com/youtrack/issue/DSP-499)
